### ISSUE TITLE
How do you prune?

### ISSUE THREAD

**Issue by zhaoxiuzhu** [Valdecy/C4.5#3]
First of all, I apologize for not looking closely at your code, but both pre-pruning and post-pruning require validation sets for test pruning, but your Part 4 code does not have validation sets

**Comment by Valdecy** [comment]
The "pre_pruning" stops to investigate the branch if a condition is reached. The conditions are:  If "chi_2" is selected then a pre-pruning method based on a Chi Squared test is performed. If "impur" is selected then a pre-pruning method is performed, pruning child nodes that do not improve the impurity from its father node. if "min" is selected then a node must have a minimum quantity of data examples to avoid pruning.

**Comment by zhaoxiuzhu** [comment]
First of all, thank you very much for answering me, which is very encouraging for a beginner like me.Secondly, I still have two problems that I am not very clear about. I hope you can help me.
1. The parameter "impur" I selected implements the pre-pruning method, that is, pruning from the parent node does not improve the child node of the impurity, but how to determine whether the child node improves the impurity of the parent node?The book I found says to separate out validation sets in the training set for pruning, but I didn't find the contents of validation sets in your parameters.
2.   I use your C4.5 code to deal with continuous values more classification task but appear error, after changed to binary classification is ok, please your code can handle only a binary classification task, if can handle multiple classification, should be how to change 
I hope you can help me, also feel sorry for the delay your time

**Comment by Valdecy** [comment]
You are Welcome!
1) I have prepared a lesson for my classes about C4.5, I think it should clear your doubts: https://www.researchgate.net/publication/344389246_DS-Classification-04-C45
2) It can handle multiple classifications, I suggest that you divide the continuous values in ranges to represent classes (for example 1.1 - 2.0 = C1; 2.1 - 3.0 = C2; and so on)

**Comment by zhaoxiuzhu** [comment]
Thank you very much indeed

**Comment by zhaoxiuzhu** [comment]
I really appreciate your help. I have learned a lot about decision tree, but my programming ability is not very good, so there is still a problem.I want to apply your code to some other data sets in order to understand and become familiar with the process, but I'm having trouble.I chose the classic Iris algorithm from the UCI database, which is a three-label data set with four features.I copied the data set to the file python-dm-classification-04-c4.5.csv as shown in the figure below
![111](https://user-images.githubusercontent.com/77867514/116022294-8146fa00-a67c-11eb-8b8f-11cd338675e8.png)
, and then output the prediction
![5](https://user-images.githubusercontent.com/77867514/116022422-c4a16880-a67c-11eb-86cf-d6aedc27a1bb.png)
, then the following problem occurred.
![2](https://user-images.githubusercontent.com/77867514/116022326-90c64300-a67c-11eb-8551-73c8f381063e.png)
 
When I deleted the data labeled 2 in the data set, it was successfully run
![66](https://user-images.githubusercontent.com/77867514/116022988-dc2d2100-a67d-11eb-9527-bd2b9982a8f0.png)
. But I want to deal with multiple classifications，I took your advice and changed your code to the following.Your original code is
![3](https://user-images.githubusercontent.com/77867514/116022715-514c2680-a67d-11eb-94cf-e7eb97b272df.png)
, I changed is
![4](https://user-images.githubusercontent.com/77867514/116022732-57420780-a67d-11eb-8b09-04d834a59cf9.png)

But the problem is still running.I wonder if something needs to change in the code when it's applied to other data but I don't notice it? Can you help me figure out how to use your code to deal with other data sets, in the case of Iris

**Comment by Valdecy** [comment]
Hello!

Indeed there was a little bug in the code for variables with decimal numbers, and I have corrected it. Thank you very much!

Also, I have implemented the code in Google Colab, using the Iris dataset for you:

https://colab.research.google.com/drive/1upOPNWJbMEW5BOUIuy2Ubna3vOpV4NYs?usp=sharing

**Comment by zhaoxiuzhu** [comment]
Hello! 
I'm glad you helped me so much. The reason why I didn't reply you for such a long time is that my country has a little blockade on the external network, so it is a little difficult to log in Google. But today I finally saw the code you helped me to change, I also ran it, the result is similar to your presentation, but there is still a problem, and I did not see it in your code. It's prediction. I used your prediction function, but all the predicted data were 0. Below is my screenshot, the first one is the code should be OK, 
![运行代码](https://user-images.githubusercontent.com/77867514/116377761-19431000-a844-11eb-850b-1d06a51fff83.png)
the second one is my test set
![测试](https://user-images.githubusercontent.com/77867514/116377810-24963b80-a844-11eb-9bc2-b4c13c700903.png)
, and the third one is my running result
![运行结果](https://user-images.githubusercontent.com/77867514/116377842-2bbd4980-a844-11eb-9851-74b4706fcbb8.png)
. I don't know if it's my processing problem or the predictive function problem? And why does the predicted data come with “；” ? In addition, I think your prediction function will be a good choice if it could be cross-validated (such as five-fold) and display both the prediction value and the accuracy at the same time.

**Comment by Valdecy** [comment]
Hello!

I have included in the colab link, the prediction using the full data and the k-fold validation (k = 5)

**Comment by zhaoxiuzhu** [comment]
Hello! It's me. These days I have a rest because of the holiday, but I still have a problem.I feel that the advantage of your code is that it has more pruning and missing value handling and so on than other people, and I don't apply missing value handling and so on and I focus on pre-pruning. I downloaded your code from colab and applied it on wine data set. The first time I ran it, I followed your code, but I preferred to use prepruning for pruning. However, when I only changed this code of yours
![代码更改](https://user-images.githubusercontent.com/77867514/117278980-d3b1c300-ae93-11eb-9ed3-691c67a2bd2b.png)

, the following error occurred
![错误](https://user-images.githubusercontent.com/77867514/117279001-d9a7a400-ae93-11eb-8563-9961de80ce47.png)


Then I changed it to the following operation with the same error
![二改](https://user-images.githubusercontent.com/77867514/117279023-df04ee80-ae93-11eb-9fb0-7ab5b6c99df7.png)
Here is the result of your source code, is it too good or too fit, please tell me how to do pre-pruning
![好的](https://user-images.githubusercontent.com/77867514/117280153-e4af0400-ae94-11eb-81cb-2d45c94b1e77.png)

**Comment by zhaoxiuzhu** [comment]
Hello! It's me. These days I have a rest because of the holiday, but I still have a problem.I feel that the advantage of your code is that it has more pruning and missing value handling and so on than other people, and I don't apply missing value handling and so on and I focus on pre-pruning. I downloaded your code from colab and applied it on wine data set. The first time I ran it, I followed your code, but I preferred to use prepruning for pruning. However, when I only changed this code of yours
![代码更改](https://user-images.githubusercontent.com/77867514/117278980-d3b1c300-ae93-11eb-9ed3-691c67a2bd2b.png)

, the following error occurred
![错误](https://user-images.githubusercontent.com/77867514/117279001-d9a7a400-ae93-11eb-8563-9961de80ce47.png)


Then I changed it to the following operation with the same error
![二改](https://user-images.githubusercontent.com/77867514/117279023-df04ee80-ae93-11eb-9fb0-7ab5b6c99df7.png)
Here is the result of your source code, is it too good or too fit, please tell me how to do pre-pruning
![好的](https://user-images.githubusercontent.com/77867514/117280153-e4af0400-ae94-11eb-81cb-2d45c94b1e77.png)

**Comment by Valdecy** [comment]
Hi Zhao! Well, I have uploaded the wine dataset, so just change:
df = pd.read_csv('https://github.com/Valdecy/Datasets/raw/master/Data%20Science/AI-11-DT.csv', sep = ';')
to
df = pd.read_csv('https://github.com/Valdecy/Datasets/raw/master/Data%20Science/AI-12-DT.csv', sep = ';')

I have run the code ( dt_model = dt_c45(Xdata = X, ydata = y, pre_pruning = "impur") ) and all seems ok. The pre_pruning = '' impur'' have sliced the branches too much so it could not predict the Class A (1). Try pre_pruning = ''min'' for better results

**Comment by zhaoxiuzhu** [comment]
Thank you very much, my friend. I changed it with your code area without making any changes, but I made a mistake. But after I changed it on colab, I found it was normal, and I think it should be the problem of the data set. Because I'm using my own data set and maybe the data set doesn't match in some places. What should I pay attention to when using my own data set? What format should be used, what encoding should be used, etc.

**Comment by Valdecy** [comment]
You're welcome, my friend! I think the main point of attention, should be with the encoding of the Classes. Try to encode them as letters, for example: Class 1 =  A; Class 2 = B and so on.

**Comment by zhaoxiuzhu** [comment]
Thank you very much! Another question is that I have conducted  Pre_pruning = '' impur''  tests on both your IRI and WINE data sets, but I have found that the effects are not good.  Why is that?  Pre_pruning = '' impur''  was supposed to do better, but the results were poor. Is that true for all the other data sets？

**Comment by Valdecy** [comment]
Yes, I notice that. I will investigate this issue, however, I believe that will not be true for all datasets

**Comment by zhaoxiuzhu** [comment]
Yes, I think so too. And I really appreciate your help， my friend. Have a good day！

**Comment by zhaoxiuzhu** [comment]
Hello, my friend. I ran into two problems while using your code that I felt would help improve your code as well. I used your decision tree code to test the data set Winequality - White with 4898 pieces of data. Because of its high imbalance, I first oversampled the data and then the data volume reached about 14000 pieces. 
 Then when I use your decision tree to test it, after 12 hours without a result, I think this is why, **if I want to see the final result, how to improve the speed of the way?**  **And can you upload Winequality - White set as well?** Also I think using StratifiedKfold for stratified sampling instead of Kfold is a better choice。

The second question is how to set the following parameters when pre_pruning="min". Is there a quick way to search for the best parameters?

**Comment by Valdecy** [comment]
Hi, my friend! I have uploaded both datasets Red Wine Quality (AI-13-DT-Red.txt) and White Wine Quality (AI-14-DT-White.txt). First of all, as I use pandas dataframes the code runs very slow, so I have to change the code to use numpy arrays to increase the speed. However, I believe that my code will not scale well, even using numpy that's mainly the reason I did not change the code yet.  For the second question, when you are using pre_pruning="min",  set min_lim = 5 or less. I don't how to get the best parameters in a fast way, usually testing and comparing each parameter (grid search) should be enough
